File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0160.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /// O(n+m) algorithm with O(1) extra space
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode p = headA;
        ListNode q = headB;

        while (p != q){
            p = p == null? headB: p.next;
            q = q == null? headA: q.next;
        } // end-while

        return p;        
    } // end-getIntersectionNode
}

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0160.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # O(n+m) algorithm with O(1) extra space
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        p = headA
        q = headB

        while p != q:
            p = headB if p == None else p.next
            q = headA if q == None else q.next
        return p 